src/mem/address_space.h

```diff
@@ -1,10 +1,96 @@
 #include "../ds/address.h"
 #include "../ds/flaglock.h"
 #include "../pal/pal.h"
+#include "pagemap.h"
 
 #include <array>
 namespace snmalloc
 {
+  template<
+    SNMALLOC_CONCEPT(ConceptPAL) PAL,
+    SNMALLOC_CONCEPT(ConceptAAL) AAL,
+    typename PrimAlloc>
+  class AuthMap
+  {
+    static_assert(
+      !aal_supports<StrictProvenance> || pal_supports<AlignedAllocation, PAL>,
+      "StrictProvenance requires platform support for aligned allocation");
+
+    /*
+     * Compute the block allocation size to use for AlignedAllocations.  This
+     * is either AAL::ptrauth_root_alloc_size, on architectures that support
+     * StrictProvenance, or a placeholder value that just needs to appease
+     * the compiler and won't otherwise be used.
+     */
+    struct default_alloc_size_t
+    {
+      static constexpr size_t ptrauth_root_alloc_size =
+        bits::one_at_bit(bits::ADDRESS_BITS - 4);
+    };
+
+  public:
+    static constexpr size_t alloc_size = std::conditional_t<
+      aal_supports<StrictProvenance, AAL>,
+      AAL,
+      default_alloc_size_t>::ptrauth_root_alloc_size;
+
+    static_assert(
+      alloc_size % PAL::page_size == 0,
+      "Provenance root granule size must be a multiple of page size");
+
+    static_assert(
+      !aal_supports<StrictProvenance, AAL> ||
+        (bits::next_pow2_const(alloc_size) == alloc_size),
+      "Provenace root granule size must be a power of two");
+
+  private:
+    static constexpr size_t AUTHMAP_BITS =
+      bits::next_pow2_bits_const(alloc_size);
+
+    static constexpr bool AUTHMAP_USE_FLATPAGEMAP = pal_supports<LazyCommit> ||
+      (PAGEMAP_NODE_SIZE >= sizeof(FlatPagemap<AUTHMAP_BITS, void*>));
+
+    struct default_auth_pagemap
+    {
+      template<typename T>
+      static SNMALLOC_FAST_PATH AuthPtr<T> get(address_t a)
+      {
+        UNUSED(a);
+        return nullptr;
+      }
+    };
+
+    using AuthPagemap = std::conditional_t<
```

Comment:
Please add some comments for things like this - it's a complex expression that it's very easy to misread!